lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp

```diff
@@ -0,0 +1,627 @@
+
+    if (!result.portAnnos.empty())
+      modules.push_back(std::move(result));
+  }
+
+#ifndef NDEBUG
+  for (auto m : modules) {
+    LLVM_DEBUG(llvm::dbgs() << "Extmodule " << m.extModule.getName() << " has "
+                            << m.portAnnos.size() << " port annotations\n");
+  }
+#endif
+
+  // Fast path if there's nothing to do.
+  if (modules.empty()) {
+    markAllAnalysesPreserved();
+    return;
+  }
+
+  // Build a generator for absolute module and instance paths in the design.
+  InstancePaths instancePaths(getAnalysis<InstanceGraph>());
+
+  // Gather the annotated ports and operations throughout the design that we are
+  // supposed to tap in one way or another.
+  tappedArgs.clear();
+  tappedOps.clear();
+  circuitOp.walk([&](Operation *op) { gatherAnnotations(op); });
+
+#ifndef NDEBUG
+  LLVM_DEBUG(llvm::dbgs() << "Tapped values:\n");
+  for (auto it : tappedArgs)
+    LLVM_DEBUG(llvm::dbgs() << "- " << it.first << ": " << it.second << "\n");
+  LLVM_DEBUG(llvm::dbgs() << "Tapped ops:\n");
+  for (auto it : tappedOps)
+    LLVM_DEBUG(llvm::dbgs() << "- " << it.first << ": " << *it.second << "\n");
+#endif
+
+  // Process each black box independently.
+  for (auto blackBox : modules) {
+    LLVM_DEBUG(llvm::dbgs() << "Generating impls for "
+                            << blackBox.extModule.getName() << "\n");
+
+    // As a first step, gather a list of all absolute paths to instances of this
+    // black box.
+    auto paths = instancePaths.getAbsolutePaths(blackBox.extModule);
+#ifndef NDEBUG
+    for (auto path : paths)
+      LLVM_DEBUG(llvm::dbgs() << "- " << path << "\n");
+#endif
+
+    // Go through the port annotations of the tap module and generate a
+    // hierarchical path for each.
+    portWiring.clear();
+    portWiring.reserve(blackBox.portAnnos.size());
+    for (auto portAnno : blackBox.portAnnos) {
+      processAnnotation(portAnno, blackBox, instancePaths);
+    }
+
+#ifndef NDEBUG
+    LLVM_DEBUG(llvm::dbgs() << "- Wire up as follows:\n");
+    for (auto wiring : portWiring) {
+      LLVM_DEBUG(llvm::dbgs() << "- Port " << wiring.portNum << ":\n");
+      for (auto path : wiring.prefices) {
+        LLVM_DEBUG(llvm::dbgs()
+                   << "  - " << path << "." << wiring.suffix << "\n");
+      }
+    }
+#endif
+
+    // Now we have an awkward mapping problem. We have multiple data tap module
+    // instances, which reference things in modules that in turn have multiple
+    // instances. This is a side-effect of how Grand Central annotates things on
+    // modules rather than instances. (However in practice these will have a
+    // one-to-one correspondence due to CHIRRTL having fully uniquified
+    // instances.) To solve this issue, create a dedicated implementation for
+    // every data tap instance, and among the possible targets for the data taps
+    // choose the one with the shortest relative path to the data tap instance.
```

Comment:
This whole thing makes me nervous... I'm not sure what to do about it, though.

Reply:
Totally agree. One thing that could help here is if we could somehow guarantee that GCT can only tap data/mems in the same subtree. So basically, module "Foo" can only tap data/mems that are in "Foo" itself, or in one of its sub-modules. That would significantly simplify the problem, because one side of the hierarchical lookup becomes trivial (we no longer need to climb upwards in the hierarchy to find a common ancestor).